src/lib/Imp.hs

```diff
@@ -269,178 +307,311 @@ toImpHof env (maybeDest, hof) = do
+  return $ Nest (DataConRefBinding (Bind v) dest) rest''
+
+applyIdxs :: MonadEmbed m => Atom -> IndexStructure -> m Atom
+applyIdxs ptr Empty = return ptr
+applyIdxs ptr idxs@(Nest (Bind i) rest) = do
+  ordinal <- indexToIntE $ Var i
+  offset <- offsetToE idxs ordinal
+  ptr' <- ptrOffset ptr offset
+  applyIdxs ptr' rest
+
+copyAtom :: Dest -> Atom -> ImpM ()
+copyAtom (BoxedRef b ptrPtr size body) src = do
+  -- TODO: load old ptr and free (recursively)
+  let PtrTy ptrTy = binderAnn b
+  size' <- translateBlock mempty (Nothing, size)
+  ptr <- emitAlloc ptrTy $ fromScalarAtom size'
+  body' <- impSubst (b@>toScalarAtom ptr) body
+  copyAtom body' src
+  store (fromScalarAtom ptrPtr) ptr
+copyAtom (DataConRef _ _ refs) (DataCon _ _ _ vals) = copyDataConArgs refs vals
+copyAtom (Con dest) src = case (dest, src) of
+  (BaseTypeRef ptr, _) -> store (fromScalarAtom ptr) (fromScalarAtom src)
+  (TabRef _, TabVal _ _) -> zipTabDestAtom copyAtom (Con dest) src
+  (ConRef (SumAsProd _ tag payload), DataCon _ _ con x) -> do
+    copyAtom tag (TagRepVal $ fromIntegral con)
+    zipWithM_ copyAtom (payload !! con) x
+  (ConRef destCon, Con srcCon) -> zipWithRefConM copyAtom destCon srcCon
+  (RecordRef refs, Record vals)
+    | fmap (const ()) refs == fmap (const ()) vals -> do
+        zipWithM_ copyAtom (toList refs) (toList vals)
+  (ConRef (SumAsProd _ tag payload), Variant (NoExt types) label i x) -> do
+    let LabeledItems ixtypes = enumerate types
+    let index = fst $ (ixtypes M.! label) NE.!! i
+    copyAtom tag (TagRepVal $ fromIntegral index)
+    zipWithM_ copyAtom (payload !! index) [x]
+  _ -> error $ "Not implemented: " ++ pprint (dest, src)
+
+copyDataConArgs :: Nest DataConRefBinding -> [Atom] -> ImpM ()
+copyDataConArgs Empty [] = return ()
+copyDataConArgs (Nest (DataConRefBinding b ref) rest) (x:xs) = do
+  copyAtom ref x
+  rest' <- impSubst (b@>x) rest
+  copyDataConArgs rest' xs
+
+loadDest :: MonadEmbed m => Dest -> m Atom
+loadDest (BoxedRef b ptrPtr _ body) = do
+  ptr <- ptrLoad ptrPtr
+  body' <- substEmbed (b@>ptr) body
+  loadDest body'
+loadDest dest@(DataConRef def params bs) = do
+  DataCon def params 0 <$> loadDataConArgs bs
+loadDest (Con dest) = do
+ case dest of
+  -- BaseTypeRef ptr -> ptrLoad ptr
+  BaseTypeRef ptr -> ptrLoad ptr
+  TabRef (TabVal b body) -> buildLam b TabArrow $ \i -> do
+    body' <- substEmbed (b@>i) body
+    result <- emitBlock body'
+    loadDest result
+  ConRef con -> Con <$> case con of
+    PairCon d1 d2 -> PairCon <$> loadDest d1 <*> loadDest d2
+    UnitCon -> return UnitCon
+    SumAsProd ty tag xss -> SumAsProd ty <$> loadDest tag <*> mapM (mapM loadDest) xss
+    IntRangeVal     l h iRef -> IntRangeVal     l h <$> loadDest iRef
+    IndexRangeVal t l h iRef -> IndexRangeVal t l h <$> loadDest iRef
+    CharCon d -> CharCon <$> loadDest d
+  RecordRef xs -> Record <$> traverse loadDest xs
+  _ -> error $ "Not implemented: " ++ pprint dest
```

Comment:
Can we change this to "not a valid dest"? Pieces of code like this one are the only documentation we have, so it's good to be explicit about that.